### PR TITLE
Non lazy joins 3566

### PR DESCRIPTION
**Pull request by romainfrancois**
This is a follow up PR from #3574 
The idea is to get rid of lazy grouped tibbles since all verbs can now make them greedily.

### CODE REVIEW

**Comment by romainfrancois** [comment]
Should this be stronger ? 

```r
#' @export
group_data.grouped_df <- function(.data) {
  res <- attr(.data, "groups")
  if (!is.data.frame(res)) {
    bad_args(".data", "is a corrupt grouped tibble")
  }
  res
}
```

i.e. testing that the last column is called `.rows`, is a list of integer vectors ? it's a bit more expensive but it could be worth it in case users start to mess with the `.groups` attribute

**Comment by romainfrancois** [comment]
@krlmlr in this I've reduced the number of calls to the groups attribute directly. Most of it happens inside the GroupedDataFrame class. (#3953)

**Comment by krlmlr** [comment]
How about implementing `group_data.grouped_df()` in C++, by constructing a `GroupedDataFrame()` object (which should do the checks anyway) and calling the accessor? The consistency check can be made pretty fast there. That would put the responsibility of the `"groups"` attribute solely into the hands of the C++ class, too.

File: src/select.cpp [hunk 1]
```diff
@@ -32,43 +32,29 @@ SEXP select_not_grouped(const DataFrame& df, const SymbolVector& keep, const Sym
...
-    if (j.origin != NEW) {
-      vars.set(i, new_names[j.pos]);
+  // update the names of the grouping variables in case they are involved in
+  // the selection, i.e. select(data, g1 = g2)
+  CharacterVector group_names = clone<CharacterVector>(groups.names());
+  IntegerVector positions = keep.match(group_names);
+  int nl = group_names.size();
+  for (int i = 0; i < nl; i++) {
+    int pos = positions[i];
+    if (pos != NA_INTEGER) {
```
Comment (krlmlr) [inline src/select.cpp]: Do we need to throw an error in case of mismatch?

Comment (romainfrancois) [reply src/select.cpp]: no because this would have been caught much earlier by `vars_select`. I guess that test is useless.

**Comment by romainfrancois** [comment]
@krlmlr not sure I follow the problem you describe re serialization. it's just a data frame that has another data frame as an attribute. R serialization can cope with that. 

Anyway, all uses of attr("groups") are now inside GroupedDataFrame, either in a static or instance method.

**Comment by krlmlr** [comment]
Thanks! I wonder what happens if a user has a serialized a grouped data frame with dplyr <= 0.7.5 and deserializes it with this version of dplyr.

**Comment by romainfrancois** [comment]
Ah I see what you mean. Can you open an issue, we'll continue the discussion in there. 

In the meantime, I think this is good to go.

File: NAMESPACE [hunk 2]
```diff
@@ -455,6 +455,7 @@ importFrom(tibble,type_sum)
 importFrom(tidyselect,contains)
 importFrom(tidyselect,ends_with)
 importFrom(tidyselect,everything)
+importFrom(tidyselect,last_col)
```
Comment (krlmlr) [inline NAMESPACE]: I remember filing an issue, a NEWS entry seems useful. (This was an unsolved question in my recent workshop.)

File: src/group_indices.cpp [hunk 3]
```diff
@@ -477,34 +477,39 @@ SEXP build_index_cpp(const DataFrame& data, const SymbolVector& vars) {
...
 
-  if (Rf_isNull(groups)) {
+  // groups must be a data frame
+  if (!is<DataFrame>(groups)) {
+    bad_arg(".data", "is a corrupt grouped_df");
+  }
+
+  // it must have at least 1 column
+  int nc = Rf_length(groups);
+  if (nc == 0) {
```
Comment (krlmlr) [inline src/group_indices.cpp]: `nc >= 1` maybe?

File: src/select.cpp (hunk 1 above)
Comment (krlmlr) [reply src/select.cpp]: The test saves us from a segmentation fault if for some reason the group variables are omitted. Can we just `stop()` if this test hits?

File: NAMESPACE (hunk 2 above)
Comment (romainfrancois) [reply NAMESPACE]: I don't know what you refer to here.

File: src/group_indices.cpp (hunk 3 above)
Comment (romainfrancois) [reply src/group_indices.cpp]: ✅

File: src/select.cpp (hunk 1 above)
Comment (romainfrancois) [reply src/select.cpp]: Do you have code that used to trigger these segfaults. I don't see how pos can ever be NA here.

File: NAMESPACE (hunk 2 above)
Comment (krlmlr) [reply NAMESPACE]: This closes #3584, I'd think a NEWS entry might be useful here.

File: src/select.cpp (hunk 1 above)
Comment (krlmlr) [reply src/select.cpp]: I don't have code now. But this code is at the mercy of two other components (R code in *dplyr*, and *tidyselect*) working correctly. If they for some reason fail, a `stop()` would give us a very clear indication about the problem. A segfault is just as clear (for the developer), but it tends to scare users.

Also, doesn't `group_data()` contain `.rows` as the last column?

File: NAMESPACE (hunk 2 above)
Comment (romainfrancois) [reply NAMESPACE]: ✅

File: src/select.cpp (hunk 1 above)
Comment (romainfrancois) [reply src/select.cpp]: There actually was a test which I had to update for this new error: 

```
test_that("rename errors with invalid grouped data frame (#640)", {
  df <- data_frame(a = 1:3, b = 2:4, d = 3:5) %>% group_by(a, b)
  df$a <- NULL
  expect_error(
    df %>% rename(e = d),
    "not found in groups metadata"
  )
  expect_error(
    df %>% rename(e = b),
    "not found in groups metadata"
  )
})
```

Comment (romainfrancois) [reply src/select.cpp]: 🤦‍♂️ except that I pushed it to the wrong branch `nest_join-3570` that depends on this one. I'll just merge this one now and the other one quickly after. Need more Monday ☕️.

**Pull request merged**